x-pack/plugin/esql/src/internalClusterTest/java/org/elasticsearch/xpack/esql/action/CrossClusterQueryWithPartialResultsIT.java

```diff
@@ -0,0 +1,394 @@
+import static org.elasticsearch.xpack.esql.EsqlTestUtils.getValuesList;
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.greaterThan;
+import static org.hamcrest.Matchers.in;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.lessThanOrEqualTo;
+
+public class CrossClusterQueryWithPartialResultsIT extends AbstractCrossClusterTestCase {
+    private static final AtomicLong nextDocId = new AtomicLong(0);
+
+    private static class ClusterSetup {
+        final int okShards = randomIntBetween(1, 5);
+        final int failingShards = randomIntBetween(1, 5);
+        Set<String> okIds;
+    }
+
+    private final ClusterSetup local = new ClusterSetup();
+    private final ClusterSetup remote1 = new ClusterSetup();
+    private final ClusterSetup remote2 = new ClusterSetup();
+
+    void populateIndices() throws Exception {
+        local.okIds = populateIndex(LOCAL_CLUSTER, "ok-local", local.okShards);
+        populateIndexWithFailingFields(LOCAL_CLUSTER, "fail-local", local.failingShards);
+
+        remote1.okIds = populateIndex(REMOTE_CLUSTER_1, "ok-cluster1", remote1.okShards);
+        populateIndexWithFailingFields(REMOTE_CLUSTER_1, "fail-cluster1", remote1.failingShards);
+
+        remote2.okIds = populateIndex(REMOTE_CLUSTER_2, "ok-cluster2", remote2.okShards);
+        populateIndexWithFailingFields(REMOTE_CLUSTER_2, "fail-cluster2", remote2.failingShards);
+    }
+
+    public void testPartialResults() throws Exception {
+        populateIndices();
+        EsqlQueryRequest request = new EsqlQueryRequest();
+        request.query("FROM ok*,fail*,*:ok*,*:fail* | KEEP id, fail_me | LIMIT 1000");
+        request.includeCCSMetadata(randomBoolean());
+        {
+            request.allowPartialResults(false);
+            IllegalStateException error = expectThrows(IllegalStateException.class, () -> runQuery(request).close());
+            assertThat(error.getMessage(), containsString("Accessing failing field"));
+        }
+        request.allowPartialResults(true);
+        try (var resp = runQuery(request)) {
+            assertTrue(resp.isPartial());
+            Set<String> allIds = Stream.of(local.okIds, remote1.okIds, remote2.okIds)
+                .flatMap(Collection::stream)
+                .collect(Collectors.toSet());
+            List<List<Object>> rows = getValuesList(resp);
+            assertThat(rows.size(), lessThanOrEqualTo(allIds.size()));
+            Set<String> returnedIds = new HashSet<>();
+            for (List<Object> row : rows) {
+                assertThat(row.size(), equalTo(2));
+                String id = (String) row.get(0);
+                assertTrue(returnedIds.add(id));
+                assertThat(id, is(in(allIds)));
+            }
+            if (request.includeCCSMetadata()) {
+                EsqlExecutionInfo.Cluster localInfo = resp.getExecutionInfo().getCluster(LOCAL_CLUSTER);
```

Comment:
This probably should be a common method, but we can refactor it in a later patch.